core/src/machine/input.rs

```diff
@@ -0,0 +1,211 @@
+use bit_field::BitField;
+
+use crate::{
+    primitives::Byte,
+    env::Input,
+    machine::interrupt::InterruptController,
+};
+
+
+/// Manages the input from the Joypad. This is mapped to 0xFF00 in the Memory.
+pub(crate) struct InputController {
+    // TODO: Implement Joypad Interrupt
+    register: Byte,
+}
+
+impl InputController {
+    /// Creates an instance with no buttons pressed and no input selected.
+    pub(crate) fn new() -> Self {
+        Self {
+            register: Byte::new(0xFF),
+        }
+    }
+
+    /// Loads the input register.
+    ///
+    /// This function behaves like the real input register. Meaning: Bits 6 and 7 always return
+    /// 1.
+    pub(crate) fn load_register(&self) -> Byte {
+        self.register.map(|b| b | 0b1100_0000)
+    }
+
+    /// Stores a byte to the input register.
+    ///
+    /// This function behaves like the real input register. Meaning: Only Bits 5 and 4 are
+    /// writable.
+    pub(crate) fn store_register(&mut self, byte: Byte) {
+        let mask = 0b0011_0000;
+        self.register = byte.map(|b| b & mask) | self.register.map(|b| b & !mask);
+    }
+
+    /// Reacts to the input transmitted via the input parameter.
+    pub(crate) fn handle_input(
+        &mut self,
+        input: &Input,
+        _interrupt_controller: &mut InterruptController,
+    ) {
+        let pressed = input.get_pressed_keys();
+        let keys = match (self.is_direction_selected(), self.is_button_selected()) {
+            (false, false) => 0,
+            (false, true) => pressed.get_button_keys(),
+            (true, false) => pressed.get_direction_keys(),
+            (true, true) => pressed.get_direction_keys() | pressed.get_button_keys(),
+        };
+
+        self.register = self.register.map(|r| {
+            (r & 0b1111_0000) | (!keys & 0b0000_1111)
+        });
+    }
+
+    /// Returns true, if the button keys are selected, false otherwise.
+    fn is_button_selected(&self) -> bool {
+        (self.register.get() & 0b0010_0000) == 0
+    }
+
+    /// Returns true, if the direction keys are selected, false otherwise.
+    fn is_direction_selected(&self) -> bool {
+        (self.register.get() & 0b0001_0000) == 0
+    }
+}
+
+/// Represents the buttons pressed on the Joypad in an easy and convenient way (some people say,
+/// Nintedo should have implemented their Joypad register this way). The bits in this u8
+/// represent the buttons with 0: not pressed and 1: pressed. The relation is:
+/// 0: A    (LSB)
+/// 1: B
+/// 2: Select
+/// 3: Start
+/// 4: Right
+/// 5: Left
+/// 6: Up
+/// 7: Down (MSB)
```

Comment:
Add `-` bullet points at the start of each line. Then it's gonna be a real list, even after the markdown doc renderer is finished with it.